libs/@local/hashql/core/src/module/std_lib.rs

```diff
@@ -0,0 +1,395 @@
+            name: self.heap.intern_symbol("::kernel::type::Some"),
+            repr: self.alloc_type(TypeKind::Param(Param {
+                argument: some_generic,
+            })),
+            arguments: self.env.intern_generic_arguments(&mut [GenericArgument {
+                id: some_generic,
+                name: self.heap.intern_symbol("T"),
+                constraint: None,
+            }]),
+        }));
+
+        let option = self.env.alloc(|id| Type {
+            id,
+            span: SpanId::SYNTHETIC,
+            kind: self.env.intern_kind(TypeKind::Union(UnionType {
+                variants: self.env.intern_type_ids(&[some, none]),
+            })),
+        });
+
+        items.extend_from_slice(&[
+            self.alloc_type_item(parent, "None", none),
+            self.alloc_intrinsic_value(parent, "::kernel::type::None", None),
+            self.alloc_type_item(parent, "Some", some),
+            self.alloc_intrinsic_value(parent, "::kernel::type::Some", None),
+            self.alloc_type_item(parent, "Option", option),
+        ]);
+    }
+
+    fn kernel_type_module_result(&self, parent: ModuleId, items: &mut Vec<ItemId>) {
+        let value_generic = self.env.counter.generic_argument.next();
+        let error_generic = self.env.counter.generic_argument.next();
+
+        let ok = self.alloc_type(TypeKind::Opaque(OpaqueType {
+            name: self.heap.intern_symbol("::kernel::type::Ok"),
+            repr: self.alloc_type(TypeKind::Param(Param {
+                argument: value_generic,
+            })),
+            arguments: self.env.intern_generic_arguments(&mut [GenericArgument {
+                id: value_generic,
+                name: self.heap.intern_symbol("T"),
+                constraint: None,
+            }]),
+        }));
+
+        let err = self.alloc_type(TypeKind::Opaque(OpaqueType {
+            name: self.heap.intern_symbol("::kernel::type::Err"),
+            repr: self.alloc_type(TypeKind::Param(Param {
+                argument: error_generic,
+            })),
+            arguments: self.env.intern_generic_arguments(&mut [GenericArgument {
+                id: error_generic,
+                name: self.heap.intern_symbol("E"),
+                constraint: None,
+            }]),
+        }));
+
+        let result = self.env.alloc(|id| Type {
+            id,
+            span: SpanId::SYNTHETIC,
+            kind: self.env.intern_kind(TypeKind::Union(UnionType {
+                variants: self.env.intern_type_ids(&[ok, err]),
+            })),
+        });
+
+        items.extend_from_slice(&[
+            self.alloc_type_item(parent, "Ok", ok),
+            self.alloc_intrinsic_value(parent, "::kernel::type::Ok", None),
+            self.alloc_type_item(parent, "Err", err),
+            self.alloc_intrinsic_value(parent, "::kernel::type::Err", None),
+            self.alloc_type_item(parent, "Result", result),
+        ]);
+    }
```

Comment:
I'd like to show special attention to this snippet. It shows how we can represent unions (and types like `Result` easily as simply two newtypes that are unioned.

```
newtype Ok<T> = T in
newtype Err<E> = E in
type Result<T, E> = Ok<T> | Err<E>
```